### ISSUE TITLE
[PowerSyncRemote] Authorization Error with Supabase + AWS Cognito Integration: JWT Key Not Found in Keystore

### ISSUE THREAD

**Issue by rcasari1987** [powersync-ja/powersync-js#401]
## Issue Description
I'm encountering an authorization error when trying to use PowerSync with Supabase and AWS Cognito as the third-party authentication provider. The error indicates that the JWT key cannot be found in the keystore.

### Error Message
[PowerSyncRemote] [Error: Closed. Original cause [Error: [AUTHORIZATION] Authorization failed
Could not find an appropriate key in the keystore. The key is missing or no key matched the token KID].]

### Supabase Client Configuration
```typescript
import { createClient } from "@supabase/supabase-js";
import { fetchAuthSession } from "aws-amplify/auth";

export const supabaseClient = createClient(
  process.env.EXPO_PUBLIC_SUPABASE_URL!,
  process.env.EXPO_PUBLIC_SUPABASE_ANON_KEY!,
  {
    accessToken: async () => {
      const session = await fetchAuthSession();
      return session.tokens?.idToken?.toString() ?? "";
    },
  }
);
```

### PowerSync Connector Implementation
```typescript
import {
  AbstractPowerSyncDatabase,
  CrudEntry,
  PowerSyncBackendConnector,
  UpdateType,
} from "@powersync/react-native";

import { SupabaseClient } from "@supabase/supabase-js";
import { supabaseClient } from "./supabase";
import { fetchAuthSession } from "aws-amplify/auth";

/// Postgres Response codes that we cannot recover from by retrying.
const FATAL_RESPONSE_CODES = [
  // Class 22 — Data Exception
  // Examples include data type mismatch.
  new RegExp("^22...$"),
  // Class 23 — Integrity Constraint Violation.
  // Examples include NOT NULL, FOREIGN KEY and UNIQUE violations.
  new RegExp("^23...$"),
  // INSUFFICIENT PRIVILEGE - typically a row-level security violation
  new RegExp("^42501$"),
];

export class Connector implements PowerSyncBackendConnector {
  client: SupabaseClient;

  constructor() {
    this.client = supabaseClient;
  }

  async fetchCredentials() {
    const session = await fetchAuthSession();

    if (!session) {
      throw new Error(`Could not fetch credentials`);
    }

    console.debug(session.tokens);

    console.debug("session expires at", session.tokens?.idToken?.payload?.exp);

    return {
      endpoint: process.env.EXPO_PUBLIC_POWERSYNC_URL!,
      token:
        session.tokens?.idToken?.toString() ??
        "",
      expiresAt: session.tokens?.idToken?.payload?.exp
        ? new Date(session.tokens?.idToken?.payload?.exp * 1000)
        : undefined,
      userID: session.tokens?.idToken?.payload?.sub,
    };
  }

  async uploadData(database: AbstractPowerSyncDatabase): Promise<void> {
    const transaction = await database.getNextCrudTransaction();

    if (!transaction) {
      return;
    }

    let lastOp: CrudEntry | null = null;
    try {
      // Note: If transactional consistency is important, use database functions
      // or edge functions to process the entire transaction in a single call.
      for (const op of transaction.crud) {
        lastOp = op;
        const table = this.client.from(op.table);
        let result: any = null;
        switch (op.op) {
          case UpdateType.PUT:
            // eslint-disable-next-line no-case-declarations
            const record = { ...op.opData, id: op.id };
            result = await table.upsert(record);
            break;
          case UpdateType.PATCH:
            result = await table.update(op.opData).eq("id", op.id);
            break;
          case UpdateType.DELETE:
            result = await table.delete().eq("id", op.id);
            break;
        }

        if (result.error) {
          console.error(result.error);
          result.error.message = `Could not ${
            op.op
          } data to Supabase error: ${JSON.stringify(result)}`;
          throw result.error;
        }
      }

      await transaction.complete();
    } catch (ex: any) {
      console.debug(ex);
      if (
        typeof ex.code == "string" &&
        FATAL_RESPONSE_CODES.some((regex) => regex.test(ex.code))
      ) {
        /**
         * Instead of blocking the queue with these errors,
         * discard the (rest of the) transaction.
         *
         * Note that these errors typically indicate a bug in the application.
         * If protecting against data loss is important, save the failing records
         * elsewhere instead of discarding, and/or notify the user.
         */
        console.error("Data upload error - discarding:", lastOp, ex);
        await transaction.complete();
      } else {
        // Error may be retryable - e.g. network error or temporary server error.
        // Throwing an error here causes this call to be retried after a delay.
        throw ex;
      }
    }
  }
}

```
### PowerSync Provider Setup
```typescript
import "@azure/core-asynciterator-polyfill";
import { PowerSyncContext, PowerSyncDatabase } from "@powersync/react-native";
import Logger from "js-logger";
import { ReactNode, useEffect, useMemo } from "react";

import { Connector } from "@/configs/powersync-config/connector";
import { AppSchema } from "@/configs/powersync-config/schema";

Logger.useDefaults();

const connector = new Connector();

export default function PowerSyncProvider({
  children,
}: {
  children: ReactNode;
}) {
  const powerSync = useMemo(() => {
    const powerSync = new PowerSyncDatabase({
      schema: AppSchema,
      database: {
        dbFilename: "sqlite.db",
      },
    });
    powerSync.init();
    return powerSync;
  }, []);

  useEffect(() => {
    powerSync
      .connect(connector)
      .then(() => console.log("connected"))
      .catch(console.error);
  }, [powerSync]);

  return (
    <PowerSyncContext.Provider value={powerSync}>
      {children}
    </PowerSyncContext.Provider>
  );
}
```
### Expected Behavior

PowerSync should successfully authenticate using the JWT token obtained from AWS Cognito through Supabase.

### Current Behavior

Authorization is failing due to missing or mismatched JWT key in the keystore.

### Questions

Is there any additional configuration needed for PowerSync to work with AWS Cognito tokens?

Does the JWT secret in Supabase need to be configured differently when using third-party authentication?
Are there any known compatibility issues between PowerSync and AWS Cognito tokens?

Any guidance on resolving this JWT authorization issue would be greatly appreciated.

**Comment by amjadbouhouch** [comment]
@rcasari1987 any luck regarding this issue ?

**Comment by stevensJourney** [comment]
Hi @rcasari1987 , have you configured the PowerSync service with the JWKS URI for your AWS Cognito service?

The PowerSync service needs access to the public key in order to verify signed JWTs.

According to the AWS Cognito docs, a JWKS URL is of the form `https://cognito-idp.region.amazonaws.com/userPoolId/.well-known/jwks.json`

The JWKS URI can be supplied in the `Client Auth` section of the PowerSync dashboard or directly in the YAML config file if self hosting.

**Comment by rcasari1987** [comment]
Thank you for your response! Initially, I had only configured the authentication through Supabase and didn't realize I needed additional configuration in PowerSync. Following your suggestion, I edited my PowerSync instance to add the JWKS URL.
At first, I encountered an error:

[Error: Closed. Original cause [Error: [AUTHORIZATION] Authorization failed
unexpected "aud" claim value].]

I resolved this by adding the JWT audience ('aud' claim) to my instance configuration. After making this change, everything started working correctly. Thanks for pointing me in the right direction!"
